apps/src/music/views/SoundsPanel2.tsx

```diff
@@ -0,0 +1,487 @@
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import classNames from 'classnames';
+import {getBaseAssetUrl} from '../appConfig';
+import styles from './soundsPanel2.module.scss';
+import FontAwesome from '@cdo/apps/templates/FontAwesome';
+import FontAwesomeV6Icon from '@cdo/apps/componentLibrary/fontAwesomeV6Icon/FontAwesomeV6Icon';
+import MusicLibrary, {
+  SoundData,
+  SoundFolder,
+  SoundType,
+} from '../player/MusicLibrary';
+import SoundStyle from '../utils/SoundStyle';
+import FocusLock from 'react-focus-lock';
+import SegmentedButtons from '@cdo/apps/componentLibrary/segmentedButtons';
+import musicI18n from '../locale';
+
+/*
+ * Renders a UI for previewing and choosing samples. This is currently used within a
+ * custom Blockly Field {@link FieldSounds}
+ */
+
+type Mode = 'packs' | 'sounds';
+type Filter = 'all' | SoundType;
+
+type SoundEntry = {
+  folder: SoundFolder;
+  sound: SoundData;
+};
+
+// FolderPanelRow: A single folder.
+
+interface FolderPanelRowProps {
+  libraryGroupPath: string;
+  playingPreview: string;
+  folder: SoundFolder;
+  currentValue: SoundFolder;
+  onSelect: (path: SoundFolder) => void;
+  onPreview: (path: string) => void;
+  currentFolderRefCallback: (ref: HTMLDivElement) => void;
+}
+
+const FolderPanelRow: React.FunctionComponent<FolderPanelRowProps> = ({
+  libraryGroupPath,
+  playingPreview,
+  folder,
+  currentValue,
+  onSelect,
+  onPreview,
+  currentFolderRefCallback,
+}) => {
+  const previewSound = folder.sounds.find(sound => sound.type === 'preview');
+  const soundPath = previewSound && folder.id + '/' + previewSound.src;
+  const isPlayingPreview = previewSound && playingPreview === soundPath;
+  const imageSrc =
+    folder.imageSrc &&
+    `${getBaseAssetUrl()}${libraryGroupPath}/${folder.path}/${folder.imageSrc}`;
+
+  const isSelected = folder.id === currentValue.id;
+
+  const onPreviewClick = useCallback(
+    (e: Event) => {
+      if (soundPath && !isPlayingPreview) {
+        onPreview(soundPath);
+      }
+      e.stopPropagation();
+    },
+    [isPlayingPreview, onPreview, soundPath]
+  );
+
+  return (
+    <div
+      className={classNames(
+        'sounds-panel-folder-row',
+        classNames(styles.folderRow, isSelected && styles.folderRowSelected)
+      )}
+      onClick={() => onSelect(folder)}
+      onKeyDown={event => {
+        if (event.key === 'Enter') {
+          onSelect(folder);
+        }
+      }}
```

Comment:
probably should be memoized.

React 19 can't get here fast enough. 😝

Reply:
Presumably this means `useCallback`?  Done.